refactor: optimize calls to `split` and `slice` in the version.ts files

Julien-Marcou opened the pull request:
Reduce the number of calls made to `split` and `slice` in the version.ts files by storing the result the first time.

## PR Checklist

- [x] The commit message follows our guidelines: [CONTRIBUTING.md](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit)

## PR Type

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

When trying to commit, I got a warning about the `refactor` commit needing the `<scope>` field, but the [CONTRIBUTING.md](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#scope) documentation says it can be empty :

> none/empty string: useful for test and refactor changes that are done across all packages

So I'm not sure if I did it correctly.

AndrewKushnir commented on packages/animations/src/version.ts:
```diff
@@ -24,9 +24,9 @@ export class Version {
-    this.major = splits[0];
-    this.minor = splits[1];
-    this.patch = splits.slice(2).join('.');
+    this.major = splits.shift()!;
+    this.minor = splits.shift()!;
+    this.patch = splits.join('.');
```
May be we can refactor the code like this?
```suggestion
    const [major, minor, ...rest] = full.split('.');
    this.major = major;
    this.minor = minor;
    this.patch = rest.join('.');
```

alfaproject replied:
This is a bit uglier but less bytes:
```ts
    let rest: string[];
    [this.major, this.minor, ...rest] = full.split('.');
    this.patch = rest.join('.');
```

Or if RegEx is not a problem:
```ts
    [, this.major, this.minor, this.patch] = full.match(/^(\d+)\.(\d+)\.(.+)/)!;
```

Julien-Marcou replied:
I also thought about using array destructuring, but I was afraid it could be slower, so I made a little benchmark :

<details>
  <summary>Code used for the benchmark</summary>

  Tested on an old `i7 4712HQ` cpu

  ```typescript
  class Version {
    public major: string;
    public minor: string;
    public patch: string;

    constructor(public full: string) {
      const numberOfTests = 100000000;

      console.time('split');
      for (let i = 0; i < numberOfTests; i++) {
        this.versionSplit(full);
      }
      console.timeEnd('split');

      console.time('destructuring split');
      for (let i = 0; i < numberOfTests; i++) {
        this.versionDestructuringSplit(full);
      }
      console.timeEnd('destructuring split');

      console.time('condensed destructuring split');
      for (let i = 0; i < numberOfTests; i++) {
        this.versionCondensedDestructuringSplit(full);
      }
      console.timeEnd('condensed destructuring split');

      console.time('regex');
      for (let i = 0; i < numberOfTests; i++) {
        this.versionRegex(full);
      }
      console.timeEnd('regex'),

      console.time('condensed regex');
      for (let i = 0; i < numberOfTests; i++) {
        this.versionCondensedRegex(full);
      }
      console.timeEnd('condensed regex');
    }

    versionSplit(full: string) {
      const splits = full.split('.');
      this.major = splits.shift()!;
      this.minor = splits.shift()!;
      this.patch = splits.join('.');
    }

    versionDestructuringSplit(full: string) {
      const [major, minor, ...rest] = full.split('.');
      this.major = major;
      this.minor = minor;
      this.patch = rest.join('.');
    }

    versionCondensedDestructuringSplit(full: string) {
      let rest: string[];
      [this.major, this.minor, ...rest] = full.split('.');
      this.patch = rest.join('.');
    }

    versionRegex(full: string) {
      const match = full.match(/^(\d+)\.(\d+)\.(.+)/)!;
      this.major = match[1];
      this.minor = match[2];
      this.patch = match[3];
    }

    versionCondensedRegex(full: string) {
      [, this.major, this.minor, this.patch] = full.match(/^(\d+)\.(\d+)\.(.+)/)!;
    }
  }

  new Version('12.0.0');
  new Version('12.0.0-next.1');
  ```
</details>

I was not expecting the RegEx solution to be the fastest ^^

Here are the results I got using Node.js v15 (Node.js v12 gave similar results) :

|                                 | `12.0.0` + ES6 | `12.0.0-next.1` + ES6 | `12.0.0` + ES5 | `12.0.0-next.1` + ES5 |
| :------------------------------ | -------------: | --------------------: | -------------: | --------------------: |
| split                           |          26.2s |                 32.9s |          30.7s |                 33.8s |
| split + destructuring           |          27.9s |                 35.3s |          41.0s |                 48.9s | 
| split + condensed destructuring |          27.5s |                 34.2s |          41.5s |                 47.8s |
| regex                           |          11.4s |                 13.2s |          12.1s |                 13.0s |
| condensed regex                 |          11.8s |                 13.6s |          25.3s |                 25.9s |

For completness, I also tested this using Firefox and altough Firefox got better result  than Node.js for `split + "12.0.0" + ES6` (17.5s instead of 26.2s), the `regex` solution was still better (12.5s)

Transpiling to ES5 basically transformed the rest operator with a `slice()` call, which I wanted to avoid in the first place by using `shift()`, and as expected, using `slice()` yielded the worst results

Array destructuring suffers from a big performance penalty in ES5, since its wrapping the line with a `__read()` call (which is why the `condensed regex` solution is 2 times slower with ES5 than ES6)

As I expect more and more people to use ES6 or newer versions of EcmaScript, the `condensed regex` solution appears to me as the best compromise between code size & performance, what do you guys think ?

alfaproject replied:
Code/size performance, yes I would have to agree

AndrewKushnir replied:
Hi @Julien-Marcou, thanks for measuring performance.

Since the code that you've updated is not on a critical path (i.e. it's not being executed frequently) and in fact it's executed just once per page load to create an instance of the `Version` class, I'd vote for more simple/readable implementation that would simplify future maintenance.

Julien-Marcou replied:
I agree, I have no preference between a RegEx or using `split`, but a RegEx is indeed harder to maintain.

Should I go for the `split + destructuring` then ? @AndrewKushnir

AndrewKushnir replied:
@Julien-Marcou I'd vote for `split + destructuring` approach 👍

gkalpak replied:
To mix things up, I'd personally vote for the nice, declarative regex (but I am known for being biased towards regexes, so that probably doesn't count 😛).

AndrewKushnir reviewed (changes requested):
@Julien-Marcou, thanks for creating the PR.

I've added a comment with a proposed refactoring, please let me know what do you think about it.

As a part of the review process I also had a look at the commit messages and have a couple proposals:

- docs: fix package name in version.ts files in different packages
- refactor(core): optimize calls to `split` and `slice` while computing version parts

Thank you.

Julien-Marcou committed:
docs: fix package name in version.ts files in different packages

Julien-Marcou committed:
refactor(core): optimize calls to `split` and `slice` while computing version parts

Reduce the number of calls made to `split` and `slice` while computing version parts by deconstructing the result.

AndrewKushnir reviewed (approved):
@Julien-Marcou thanks for updating the code 👍 

Note: there is a payload-size-related failures on CI (the payload size was actually decreased), but we can update the size tracking JSON files once other reviewers provide their feedback.

atscott reviewed (approved):
reviewed-for: fw-router

gkalpak commented on packages/compiler-cli/src/version.ts:
```diff
@@ -9,7 +9,7 @@
  * @module
  * @description
- * Entry point for all public APIs of the common package.
+ * Entry point for all public APIs of the compiler-cli package.
```
While the package name is correct now, the comment is still wrong :grin:
This file is not the entry point of all public APIs of the package. (Same for other packages.)

(Not directly related to the PR, though.)

AndrewKushnir commented:
[Presubmit](http://test/OCL:371921976:BASE:371921999:1620142955212:9105f451).

jelbourn reviewed (approved):
LGTM

Reviewed-for: global-approvers

petebacondarwin commented:
@Julien-Marcou - can you rebase and then fix the size-tracking golden files to match the new sizes? See the CI failures for details.

AndrewKushnir commented:
@petebacondarwin I've rebased this PR a few mins ago, the payload size errors were most likely preexisting.

AndrewKushnir commented:
It looks like `test_win` and `test_ivy_aot_win` CI jobs are failing with the following error:
```
ERROR: C:/users/circleci/ng/packages/compiler-cli/test/compliance/test_cases/BUILD.bazel:14:27: output 'packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_input_outputs/this_file_should_not_be_committed' was not created
ERROR: C:/users/circleci/ng/packages/compiler-cli/test/compliance/test_cases/BUILD.bazel:14:27: Action packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_input_outputs/this_file_should_not_be_committed failed: not all outputs were created or valid
```
Which looks unrelated to the changes in this PR and rebasing this PR on top of the current master branch (which doesn't have this problem) doesn't fix the issue. @petebacondarwin have you seen this error with the compliance tests before?

petebacondarwin commented:
> It looks like `test_win` and `test_ivy_aot_win` CI jobs are failing with the following error:
> 
> ```
> ERROR: C:/users/circleci/ng/packages/compiler-cli/test/compliance/test_cases/BUILD.bazel:14:27: output 'packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_input_outputs/this_file_should_not_be_committed' was not created
> ERROR: C:/users/circleci/ng/packages/compiler-cli/test/compliance/test_cases/BUILD.bazel:14:27: Action packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_input_outputs/this_file_should_not_be_committed failed: not all outputs were created or valid
> ```
> 
> Which looks unrelated to the changes in this PR and rebasing this PR on top of the current master branch (which doesn't have this problem) doesn't fix the issue. @petebacondarwin have you seen this error with the compliance tests before?

I am confident that these are flakes. Rerunning again.

AndrewKushnir commented:
@petebacondarwin thanks for looking into that. I've tried rerunning them a couple times (with the same outcome), but may be I was not lucky enough! :)

petebacondarwin commented:
Hmm, no joy for me either... I'll look more deeply.

petebacondarwin commented:
I note that it appears to be using a remote cache:

```
build' options: --local_ram_resources=120000 --local_cpu_resources=32 --remote_http_cache=https://storage.googleapis.com/angular-team-cache --remote_accept_cached=true --remote_upload_local_results=true --google_default_credentials
```

I wonder if that cache is broken but is being reused on each run? Is there a way to invalidate that remote cache?

AndrewKushnir commented:
@petebacondarwin it looks like it was a caching issue - everything works ok after one more rebase (presumably the cache was purged at some point). I've started a [new presubmit](http://test/OCL:371921976:BASE:372727716:1620495103596:9659c03f) and this PR should be ready for merge once presubmit results are available.
